### PR TITLE
Precommit with unit tests

### PR DESCRIPTION
**Pull request by vantihovich**

### CODE REVIEW

File: swagger/configuration/configs_test.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+func MustGet() JWTParameters {
+	i, err := LoadJWT()
+	if err != nil {
+		panic(err)
+	}
+	return i
+}
+
+func TestLoadJWT(t *testing.T) {
+	assert := assert.New(t)
```
Comment (apaliavy) [inline swagger/configuration/configs_test.go:18]: Why do you need it ?

File: swagger/configuration/configs_test.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+	var tests = []struct {
+		P1 string
+		P2 int
+	}{
+		{P1: "my_secret_key", P2: 15},
```
Comment (apaliavy) [inline swagger/configuration/configs_test.go:24]: What is the purpose of looping over slice with single element?

File: swagger/configuration/configs_test.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package config
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func MustGet() JWTParameters {
```
Comment (apaliavy) [inline swagger/configuration/configs_test.go:9]: `MustGet` is not descriptive

File: swagger/configuration/configs_test.go [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	i, err := LoadJWT()
+	if err != nil {
+		panic(err)
```
Comment (apaliavy) [inline swagger/configuration/configs_test.go:12]: Don't panic

File: swagger/configuration/configs_test.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+	var tests = []struct {
+		P1 string
+		P2 int
+	}{
+		{P1: "my_secret_key", P2: 15},
+	}
+
+	for _, tt := range tests {
+		assert.Equal(tt.P2, MustGet().JwtTTL, "check TTL")
+		assert.Equal(tt.P1, MustGet().SecretKey, "check secret")
```
Comment (apaliavy) [inline swagger/configuration/configs_test.go:29]: Why do you need to call MustGet() twice? 
You can do something like 
```
cfg := MustGet()
asssert... 
```

File: swagger/gmailsmtp/smtp_test.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+	cfgJWT, err := cnfg.LoadJWT()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to load JWT config")
+	}
+
+	var tests = []struct {
+		receiver string
+	}{
+		{receiver: "vadim.antikhovich@gmail.com"},
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp_test.go:29]: Please format it properly

File: swagger/gmailsmtp/smtp_test.go [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/stretchr/testify/assert"
+
+	cnfg "github.com/vantihovich/go_tasks/tree/master/swagger/configuration"
+)
+
+func TestSendForgottenPasswordEmail(t *testing.T) {
+	assert := assert.New(t)
+
+	cfgSmtp, err := cnfg.LoadSmtpParameters()
+	if err != nil {
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp_test.go:16]: `require.NoError(t, err, "message here")`

File: swagger/gmailsmtp/smtp_test.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	}
+
+	var tests = []struct {
+		receiver string
+	}{
+		{receiver: "vadim.antikhovich@gmail.com"},
+		{receiver: "moloko31@tut.by"},
+	}
+	for _, t := range tests {
+		assert.NoError(smtp.SendForgottenPasswordEmail(t.receiver, cfgJWT.SecretKey), "check no error when sending email")
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp_test.go:33]: Does it mean that you are sending real emails from your tests? 🙀

**Review by apaliavy — CHANGES_REQUESTED** [review]

File: swagger/gmailsmtp/smtp_test.go [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	assert := assert.New(t)
+
+	cfgSmtp, err := cnfg.LoadSmtpParameters()
+	if err != nil {
+		log.WithError(err).Fatal("Failed to load SMTP configs")
+	}
+	smtp := Auth(cfgSmtp)
+
+	cfgJWT, err := cnfg.LoadJWT()
+	if err != nil {
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp_test.go:22]: require.NoError(t, err, "message here")

File: swagger/gmailsmtp/smtp_test.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+		log.WithError(err).Fatal("Failed to load JWT config")
+	}
+
+	var tests = []struct {
+		receiver string
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp_test.go:27]: If you wanna try to write table tests, I'd expect to see here some additional fields like: 
- `input`
- `expectedResult`
- `expectError`
- `expectedError`
So underlying test function uses provided input and can properly handle the result. 

In this case I don't see any reasons to use this construction. Could you please explain, from your point of view?

File: swagger/jwtgeneration/jwtgeneration_test.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/stretchr/testify/assert"
+
+	cnfg "github.com/vantihovich/go_tasks/tree/master/swagger/configuration"
+)
+
+func TestGetToken(t *testing.T) {
+	assert := assert.New(t)
+
+	cfgJWT, err := cnfg.LoadJWT()
+	if err != nil {
```
Comment (apaliavy) [inline swagger/jwtgeneration/jwtgeneration_test.go:18]: require.NoError(t, err, "message here")

File: swagger/configuration/configs_test.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
+package config
```
Comment (apaliavy) [inline swagger/configuration/configs_test.go:1]: We often use `_test` suffix here, to test only public package interface

File: swagger/mailjet/mailjet_test.go [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/stretchr/testify/assert"
+
+	cnfg "github.com/vantihovich/go_tasks/tree/master/swagger/configuration"
+)
+
+func TestSendForgottenPasswordEmail(t *testing.T) {
+	assert := assert.New(t)
+
+	cfgMailJet, err := cnfg.LoadMailJetParameters()
+	if err != nil {
```
Comment (apaliavy) [inline swagger/mailjet/mailjet_test.go:16]: require.NoError(t, err, "message here")

File: swagger/configuration/configs_test.go (hunk 1 above)
Comment (vantihovich) [reply swagger/configuration/configs_test.go:18]: removed

File: swagger/configuration/configs_test.go (hunk 2 above)
Comment (vantihovich) [reply swagger/configuration/configs_test.go:24]: removed loop

File: swagger/configuration/configs_test.go (hunk 3 above)
Comment (vantihovich) [reply swagger/configuration/configs_test.go:9]: renamed

File: swagger/configuration/configs_test.go (hunk 4 above)
Comment (vantihovich) [reply swagger/configuration/configs_test.go:12]: Yes, Sir !!! 
No panic, Sir !!!
)))

File: swagger/configuration/configs_test.go (hunk 5 above)
Comment (vantihovich) [reply swagger/configuration/configs_test.go:29]: done

File: swagger/gmailsmtp/smtp_test.go (hunk 6 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp_test.go:29]: not actual after amendment

File: swagger/gmailsmtp/smtp_test.go (hunk 7 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp_test.go:16]: done

File: swagger/gmailsmtp/smtp_test.go (hunk 8 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp_test.go:33]: after amendments smtp server is mocked and doesn`t touch real emails

File: swagger/gmailsmtp/smtp_test.go (hunk 9 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp_test.go:22]: done

File: swagger/gmailsmtp/smtp_test.go (hunk 10 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp_test.go:27]: Don`t see any reasons either. Thanks for clarifying table tests

File: swagger/jwtgeneration/jwtgeneration_test.go [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+		log.WithError(err).Fatal("Failed to load JWT config")
+	}
+
+	var tests = []struct {
+		userID int
+	}{
+		{23},
```
[the remark this answers is not in the document]
Comment (vantihovich) [reply swagger/jwtgeneration/jwtgeneration_test.go:25]: removed table test, made test for validating JWT for one userID only

File: swagger/configuration/configs_test.go (hunk 12 above)
Comment (vantihovich) [reply swagger/configuration/configs_test.go:1]: added to package name "_test"

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 11 above)
Comment (vantihovich) [reply swagger/jwtgeneration/jwtgeneration_test.go:18]: done

File: swagger/mailjet/mailjet_test.go (hunk 13 above)
Comment (vantihovich) [reply swagger/mailjet/mailjet_test.go:16]: not actual after amendments,  also mailjet service is mocked and doesn`t touch real emails .

File: swagger/gmailsmtp/smtp_test.go [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+		Sender: "testblabla@mail.com",
+	}
+
+	assert.NoError(t, mockedSmtpClient.SendForgottenPasswordEmail(receiver, cfgJWT.SecretKey), "check email sending")
+}
+
+func testServer(t *testing.T) *smtpmock.Server {
+	server := smtpmock.New(smtpmock.ConfigurationAttr{})
+	if err := server.Start(); err != nil {
+		require.NoError(t, err, "failed to start mocking smtp")
```
Comment (apaliavy) [inline swagger/gmailsmtp/smtp_test.go:47]: if err != nil => then require.NoError? Something is wrong here :D

File: swagger/jwtgeneration/jwtgeneration_test.go [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+func TestGetToken(t *testing.T) {
+	cfgJWT, err := cnfg.LoadJWT()
+	if err != nil {
+		require.NoError(t, err, "failed to load JWT configs")
+	}
+
+	userID := 23
+	token, err := jwtGeneration.GetToken(userID, cfgJWT)
+	if err != nil {
```
Comment (apaliavy) [inline swagger/jwtgeneration/jwtgeneration_test.go:24]: Same strange thing again

File: swagger/jwtgeneration/jwtgeneration_test.go [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+	jwtGeneration "github.com/vantihovich/go_tasks/tree/master/swagger/jwtgeneration"
+)
+
+func TestGetToken(t *testing.T) {
+	cfgJWT, err := cnfg.LoadJWT()
+	if err != nil {
+		require.NoError(t, err, "failed to load JWT configs")
+	}
+
+	userID := 23
```
Comment (apaliavy) [inline swagger/jwtgeneration/jwtgeneration_test.go:22]: Where it comes from?

File: swagger/jwtgeneration/jwtgeneration_test.go [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+	token, err := jwtGeneration.GetToken(userID, cfgJWT)
+	if err != nil {
+		require.NoError(t, err, "generating token error")
+	}
+
+	assert.True(t, validate(token, cfgJWT), "check if token is valid")
+}
+
+type claims struct {
```
Comment (apaliavy) [inline swagger/jwtgeneration/jwtgeneration_test.go:31]: Why do you need to re-define it here?

File: swagger/jwtgeneration/jwtgeneration_test.go [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+	UserID int `json:"user_id"`
+	jwt.StandardClaims
+}
+
+func validate(token string, jwtParams cnfg.JWTParameters) bool {
```
Comment (apaliavy) [inline swagger/jwtgeneration/jwtgeneration_test.go:36]: Why do you have a validation function in tests?

File: swagger/mailjet/mailjet_test.go [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+
+	fakeClient := mj.Client{
+		MjClient:    NewMockedClient(),
+		SenderEmail: "blablatest@mail.com",
+	}
+	assert.NoError(t, fakeClient.SendForgottenPasswordEmail("testblabla@mail.com", "testSecret"), "check email sending")
+}
+
+func NewMockedClient() *mailjet.Client {
+	smtpClientMocked := mailjet.NewSMTPClientMock(true)
```
Comment (apaliavy) [inline swagger/mailjet/mailjet_test.go:39]: Where it comes from?

File: swagger/mailjet/mailjet_test.go [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+var fakeHttpInterface = &fk.FakeHTTPClientInterface{}
+
+func TestSendForgottenPasswordEmail(t *testing.T) {
+	responseBody := `
+	{"Messages": [{"Status":"success", 
+	"To":[{"Email":"testblabla@mail.com", "MessageUUID": "test", "MessageID": 1, 
+	"MessageHref": "https://api.mailjet.com/v3/REST/message/1"}], "CC":[], "Bcc":[]}]}`
+
+	response := &http.Response{
+		StatusCode: 200,
```
Comment (apaliavy) [inline swagger/mailjet/mailjet_test.go:25]: `http.StatusOK`

File: swagger/configuration/configs_test.go [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+	i, err := cnfg.LoadJWT()
+	if err != nil {
+		require.NoError(t, err, "error when trying to get JWT parameters")
+	}
+	return i
+}
+
+func TestLoadJWT(t *testing.T) {
+
+	var tests = struct {
```
[the remark this answers is not in the document]
Comment (vantihovich) [reply swagger/configuration/configs_test.go:22]: Made just two new variables to compare the results of loading with.

File: swagger/gmailsmtp/smtp_test.go (hunk 15 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp_test.go:47]: Definitely wrong) Fixed by using t.Errorf()

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 16 above)
Comment (vantihovich) [reply swagger/jwtgeneration/jwtgeneration_test.go:24]: fixed

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 18 above)
Comment (vantihovich) [reply swagger/jwtgeneration/jwtgeneration_test.go:31]: Fixed - just made it "exported" from tested package

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 19 above)
Comment (vantihovich) [reply swagger/jwtgeneration/jwtgeneration_test.go:36]: I need to know that the generated token is valid somehow. The question here is about the reusage of the existing validator?

File: swagger/mailjet/mailjet_test.go (hunk 20 above)
Comment (vantihovich) [reply swagger/mailjet/mailjet_test.go:39]: This is just a "stub" from the mailjet`s package, it`s not used anywhere, using it as  mailjet.NewClient() needs at least  two arguments(two "clients" - http and smtp).

File: swagger/mailjet/mailjet_test.go (hunk 21 above)
Comment (vantihovich) [reply swagger/mailjet/mailjet_test.go:25]: changed

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 17 above)
Comment (vantihovich) [reply swagger/jwtgeneration/jwtgeneration_test.go:22]: I need an integer for unit testing , as the package has no specific requirements for userID field I`m just using the first one that comes in mind. If the question is about using "real test userID" , reproducing the whole cycle from the point of creating the concrete userID , or retrieving some userID from database looks a bit bulky for checking just a token generation. On the over hand when retrieving the concrete userID from DB the question might appear :why exactly this user?
Please specify what  is the better solution for such case.

File: swagger/gmailsmtp/smtp_test.go (hunk 15 above)
Comment (apaliavy) [reply swagger/gmailsmtp/smtp_test.go]: You made it even worse 😓 
```
err := server.Start()
require.NoError(t, err, "failed to start server")
```

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 16 above)
Comment (apaliavy) [reply swagger/jwtgeneration/jwtgeneration_test.go:24]: see comment above

File: swagger/configuration/configs_test.go [hunk 23]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/stretchr/testify/require"
+
+	cnfg "github.com/vantihovich/go_tasks/tree/master/swagger/configuration"
+)
+
+func getParametersOnly(t *testing.T) cnfg.JWTParameters {
+	i, err := cnfg.LoadJWT()
+	if err != nil {
+		require.NoError(t, err, "error when trying to get JWT parameters")
+	}
```
Comment (diskordanz) [inline swagger/configuration/configs_test.go:16]: require.NoError asserts that a function returned no error, no need to check whether is err is nil or not

File: swagger/configuration/configs_test.go [hunk 24]
```diff
@@ -0,0 +1,29 @@
...
+
+func TestLoadJWT(t *testing.T) {
+	secretKey := "my_secret_key"
+	jwtTtl := 15
+
+	cfg := getParametersOnly(t)
+
+	assert.Equal(t, jwtTtl, cfg.JwtTTL, "check TTL")
+	assert.Equal(t, secretKey, cfg.SecretKey, "check secret")
+
```
Comment (diskordanz) [inline swagger/configuration/configs_test.go:28]: please get rid of unnecessary empty lines across the files

File: swagger/gmailsmtp/smtp_test.go [hunk 25]
```diff
@@ -0,0 +1,50 @@
...
+
+	hostAddress, portNumber := "127.0.0.1", mockedSmtpServer.PortNumber
+	address := fmt.Sprintf("%s:%d", hostAddress, portNumber)
+	timeout := time.Duration(5) * time.Second
+
+	connection, _ := net.DialTimeout("tcp", address, timeout)
+	smtp.NewClient(connection, hostAddress)
+
+	cfgJWT, err := cnfg.LoadJWT()
+	if err != nil {
```
Comment (diskordanz) [inline swagger/gmailsmtp/smtp_test.go:29]: the same with require.NoError - there is no need for `if err != nil` check

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 11 above)
Comment (diskordanz) [reply swagger/jwtgeneration/jwtgeneration_test.go:18]: just `require.NoError(t, err, "message here")`
no need for `if err != nil` check

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 17 above)
Comment (diskordanz) [reply swagger/jwtgeneration/jwtgeneration_test.go:22]: I'd suggest to rename this variable to mockUserID then

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 19 above)
Comment (diskordanz) [reply swagger/jwtgeneration/jwtgeneration_test.go:31]: not sure about the context but using token validation here you're assuming that GetToken function could return an invalid token. so from my point of view if you really need to validate the token it's better to use it inside the function itself

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 16 above)
Comment (diskordanz) [reply swagger/jwtgeneration/jwtgeneration_test.go]: please fix it

File: swagger/gmailsmtp/smtp_test.go (hunk 15 above)
Comment (vantihovich) [reply swagger/gmailsmtp/smtp_test.go]: done

File: swagger/jwtgeneration/jwtgeneration_test.go (hunk 16 above)
Comment (vantihovich) [reply swagger/jwtgeneration/jwtgeneration_test.go]: done

**Review by diskordanz — APPROVED** [review]

**Pull request merged**